### PR TITLE
[mariadb-galera] Add script preventing race condition

### PR DESCRIPTION
**Pull request by glrf**
#### Checklist

- [x] [DCO](https://github.com/appuio/charts/blob/master/CONTRIBUTING.md#sign-your-work) signed
- [x] Variables are documented in the values.yaml using the format required by [Helm-Docs](https://github.com/norwoodj/helm-docs#valuesyaml-metadata).
- [x] Title of the PR contains starts with chart name e.g. `[chart]`

### CODE REVIEW

File: appuio/mariadb-galera/templates/recreate-script.yaml [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    NAME="{{ printf "%s" (include "common.names.fullname" .) }}"
+    NAMEPACE="{{ .Release.Namespace }}"
+
+    # Check if delete is necessary
+    FOUND=$(kubectl -n $NAMEPACE get sts $NAME -o json)
+
+    SIZE="{{ .Values.persistence.size }}"
+    FOUNDSIZE=$(echo -En $FOUND | jq -r '.spec.volumeClaimTemplates[] | select(.metadata.name=="data") | .spec.resources.requests.storage')
+
+    if [[ $FOUNDSIZE != $SIZE ]]; then
```
Comment (simu) [inline appuio/mariadb-galera/templates/recreate-script.yaml:23]:
```suggestion
    if [[ "$FOUNDSIZE" != "$SIZE" ]]; then
```

File: appuio/mariadb-galera/templates/recreate-script.yaml [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+      kubectl -n $NAMEPACE delete sts $NAME --cascade=orphan --ignore-not-found --wait=true
```
Comment (simu) [inline appuio/mariadb-galera/templates/recreate-script.yaml:24]:
```suggestion
      kubectl -n "$NAMEPACE" delete sts "$NAME" --cascade=orphan --ignore-not-found --wait=true 
```

File: appuio/mariadb-galera/templates/recreate-script.yaml [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+    SIZE="{{ .Values.persistence.size }}"
+    FOUNDSIZE=$(echo -En $FOUND | jq -r '.spec.volumeClaimTemplates[] | select(.metadata.name=="data") | .spec.resources.requests.storage')
+
+    if [[ $FOUNDSIZE != $SIZE ]]; then
+      kubectl -n $NAMEPACE delete sts $NAME --cascade=orphan --ignore-not-found --wait=true 
+      # There is a consistency issue. It seems that it is not guaranteed that the helm controller sees the deletion before noticing the job completion.
+      # It is generally hard to find clear consistency guarantees of the Kubernetes API server.
+      # My hope is that an additional read forces the delete to be committed.
+      while kubectl -n $NAMEPACE get sts $NAME; do sleep 1; done
```
Comment (simu) [inline appuio/mariadb-galera/templates/recreate-script.yaml:28]:
```suggestion
      while kubectl -n "$NAMEPACE" get sts "$NAME"; do sleep 1; done
```

File: appuio/mariadb-galera/templates/recreate-script.yaml [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+  name: {{ printf "%s-sts-deleter" (include "common.names.fullname" .) }}
+  annotations:
+    "helm.sh/hook": pre-upgrade
+    "helm.sh/hook-weight": "-6"
+  labels: {{- include "common.labels.standard" . | nindent 4 }}
+data:
+  delete.sh: |
+    #!/bin/bash
+
+    NAME="{{ printf "%s" (include "common.names.fullname" .) }}"
```
Comment (simu) [inline appuio/mariadb-galera/templates/recreate-script.yaml:14]: If you want more useful output from shellcheck, you should consider using lowercase variable names for variables which are not provided to the script through the environment.

With lowercase variables, you get the following for a typo of a variable which is used elsewhere:

```
Line 7:
foundsize=$(echo -En "$fonud" | jq -r '.spec.volumeClaimTemplates[] | select(.metadata.name=="data") | .spec.resources.requests.storage')
                      ^-- SC2154: fonud is referenced but not assigned.
```

while with uppercase variables you get no output for the same typo.

Note that you always get the following warning if you misspell a variable which you only use in a single place

```
Line 5:
FOUND=$(kubectl -n "$namespace" get sts $name -o json)
^-- SC2034: FOUND appears unused. Verify use (or export if used externally).
```

**Review by simu — CHANGES_REQUESTED** [review]
Logic LGTM.

As always when writing shell scripts: Use `set -eo pipefail` and shellcheck is your best friend and helps you avoid all the subtly unintuitive behavior in shell which you're aware of but don't think about in the moment.

**Comment by glrf** [comment]
Thank's for the input. As you can tell I don't have much experience in writing shell scripts.

**Review by simu — APPROVED** [review]

**Pull request merged**